src/backtracking/rat_in_maze.rs

```diff
@@ -0,0 +1,304 @@
+            vec![true, true, true, false, true, true, true, true],
+        ], Ok(Some(vec![
+            vec![true, false, false, false, false, false, false, false],
+            vec![true, true, false, false, false, false, false, false],
+            vec![false, true, true, true, false, false, false, false],
+            vec![false, false, false, true, false, false, false, false],
+            vec![false, false, false, true, true, true, false, false],
+            vec![false, false, false, false, false, true, true, true],
+            vec![false, false, false, false, false, false, false, true],
+            vec![false, false, false, false, false, false, false, true],
+        ])),
+        maze_without_solution_4x4: 0, 0, &[
+            vec![true, false, false, false],
+            vec![true, true, false, false],
+            vec![false, false, true, false],
+            vec![false, false, false, true],
+        ], Ok(None::<Vec<Vec<bool>>>),
+        maze_with_solution_3x4: 0, 0, &[
+            vec![true, false, true, true],
+            vec![true, true, true, false],
+            vec![false, true, true, true],
+        ], Ok(Some(vec![
+            vec![true, false, false, false],
+            vec![true, true, true, false],
+            vec![false, false, true, true],
+        ])),
+        maze_without_solution_3x4: 0, 0, &[
+            vec![true, false, true, true],
+            vec![true, false, true, false],
+            vec![false, true, false, true],
+        ], Ok(None::<Vec<Vec<bool>>>),
+        improper_maze_representation: 0, 0, &[
+            vec![true],
+            vec![true, true],
+            vec![true, true, true],
+            vec![true, true, true, true]
+        ], Err(MazeError::ImproperMazeRepr),
+        out_of_bound_start: 0, 3, &[
+            vec![true, false, true],
+            vec![true, true],
+            vec![false, true, true],
+        ], Err(MazeError::OutOfBoundPos),
+        empty_maze: 0, 0, &[], Err(MazeError::EmptyMaze),
+        maze_with_single_cell: 0, 0, &[
+            vec![true],
+        ], Ok(Some(vec![vec![true]])),
+        maze_with_one_row_and_multiple_columns: 0, 0, &[
+            vec![true, false, true, true, false]
+        ], Ok(None::<Vec<Vec<bool>>>),
+        maze_with_multiple_rows_and_one_column: 0, 0, &[
+            vec![true],
+            vec![true],
+            vec![false],
+            vec![true],
+        ], Ok(None::<Vec<Vec<bool>>>),
+        maze_with_walls_surrounding_border: 0, 0, &[
+            vec![false, false, false],
+            vec![false, true, false],
+            vec![false, false, false],
+        ], Ok(None::<Vec<Vec<bool>>>),
+        maze_with_no_walls: 0, 0, &[
+            vec![true, true, true],
+            vec![true, true, true],
+            vec![true, true, true],
+        ], Ok(Some(vec![
+            vec![true, true, true],
+            vec![false, false, true],
+            vec![false, false, true],
+        ])),
```

Comment:
Do you have some good idea to _document_ why your implementation returns this solution and not the other one? I guess it would be something like: If multiple solutions exists, the function returns the first _successful path_ it discovers, based on the predefined order of considered moves, defined in `MOVES` .

That would be an overkill to handle all of the solutions.

Reply:
You're right. Handle the case that return all possible path is overkill. I have commented what is solution matrix. Solution matrix is contain one possible path, based on the order of predefined moves.

Reply:
I will add document explain it and the trivial test is current test with single cell